inngest/src/main/kotlin/com/inngest/Event.kt

```diff
@@ -9,7 +14,83 @@ data class Event(
     val v: Any? = null,
 )
 
-// data class EventAPIResponse(
-//    val ids: Array<String>,
-//    val status: String,
-// )
+/**
+ * Create an event to send to Inngest
+ */
+data class InngestEvent
+    @JvmOverloads
+    constructor(
+        val name: String,
+        val data: Any,
+        @Json(serializeNull = false)
+        val user: Any? = null,
+        @Json(serializeNull = false)
+        val id: String? = null,
+        @Json(serializeNull = false)
+        val ts: Long? = null,
+        @Json(serializeNull = false)
+        val v: String? = null,
+    )
+
+/**
+ * Construct a new Inngest Event via builder
+ */
+class InngestEventBuilder(
+    var id: String?,
+    var name: String?,
+    var data: Any?,
+    private var user: Any?,
+    private var ts: Long?,
+    private var v: String? = null,
+) {
+    fun id(id: String): InngestEventBuilder {
+        this.id = id
+        return this
+    }
+
+    fun name(name: String): InngestEventBuilder {
+        this.name = name
+        return this
+    }
+
+    fun data(data: Any): InngestEventBuilder {
```

Comment:
`name` and `data` are mandatory so why not have an `InngestEvent` with name and data for the constructor, and then chain itself instead?

Example/Reference in Rust
https://github.com/inngest/inngest-rs/blob/main/inngest/src/event.rs#L21-L49

Reply:
Also since the `InngestEvent` is a data class with optional fields, honestly don't really need a builder?

Reply:
The general idea is that we'd have a couple different ways to do things, what you recommend and a builder pattern. It was recommended to have different options for users like this. The builder pattern might be preferred for users that want to set `id` or `v` but not have to include null arguments in a constructor.

Reply:
Following up here:

- Builder pattern is pretty established in Java because IIRC Java does not natively support optional parameters (unless that's changed in a recent version). If we use `@JvmOverloads` on a Kotlin constructor with optional params, it'll generate overloaded methods for us but not all the various permutations a user might want.
- So given that we should provide a builder at all, I think Darwin's point about putting the required parameters in the constructor is pretty good. That's what the Book example from this [article on builders from the Effective Java book has](https://blogs.oracle.com/javamagazine/post/exploring-joshua-blochs-builder-design-pattern-in-java)
  ```java
  // isbn and title are required
  public Builder(String isbn, String title) {
      this.isbn = isbn;
      this.title = title;
  }

  // genre is required
  public Builder genre(Genre genre) {
      this.genre = genre;
      return this;
  }
  ```
  However looking at these builder examples from several popular Java libraries, none of them seem to put required parameters in the builder constructor. So I guess it can go either way
  - https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/util/UriComponentsBuilder.html#build()
  - https://hc.apache.org/httpcomponents-client-5.2.x/current/httpclient5/apidocs/org/apache/hc/client5/http/config/RequestConfig.Builder.html
  - https://grpc.github.io/grpc-java/javadoc/io/grpc/ServerBuilder.html
  - https://square.github.io/okhttp/3.x/okhttp/okhttp3/Request.Builder.html
- One point for Java interop if we use non nullable constructor arguments in Kotlin. They will get annotated with `org.jetbrains.annotations.NotNull` for Java and immediately throw `java.lang.NullPointerException: Parameter specified as non-null is null` if called with a null argument. Thanks @KiKoS0 for confirming this behavior.
- A minor question is whether we want to nest the builder as `InngestEvent.Builder` or keep it as `InngestEventBuilder`. As you can see from the examples of other builders I linked, this seems to be up to the preference of the library author.